### PR TITLE
truss patch two step

### PR DESCRIPTION
**Pull request by rcano-baseten**
## :rocket: What
* patches in two steps
* concepual model: https://link.excalidraw.com/l/8mVEQNkaMDa/5vknbphhBx4

### CODE REVIEW

File: truss/remote/baseten/api.py [hunk 1]
```diff
@@ -407,13 +407,41 @@ def get_model_version_by_id(self, model_version_id: str):
...
+        result = resp["data"]["stage_patch_for_draft_truss"]
+        if not result["succeeded"]:
+            logging.debug(f"Unsuccessful response: {result}")
+            return result
+
+        return self._patch_draft_truss(model_name)
+
     def patch_draft_truss(self, model_name, patch_request):
         patch = base64_encoded_json_str(patch_request.to_dict())
+        return self._patch_draft_truss(model_name, patch)
```
Comment (marius-baseten) [inline truss/remote/baseten/api.py:437]: Why do we keep this? Or is this just temporary?

**Review by marius-baseten — APPROVED** [review]

File: truss/remote/baseten/remote.py [hunk 2]
```diff
@@ -360,19 +374,39 @@ def _patch(
                 "Failed to calculate patch. Change type might not be supported.",
             )
 
-        if (
-            patch_request.prev_hash == patch_request.next_hash
-            or len(patch_request.patch_ops) == 0
-        ):
+        requires_sync = (
```
Comment (squidarth) [inline truss/remote/baseten/remote.py:377]: just so I understand the semantic, `requires_sync` means that there are unapplied patches?

Comment (squidarth) [reply truss/remote/baseten/remote.py:377]: could we name it that way?

File: truss/remote/baseten/core.py [hunk 3]
```diff
@@ -48,6 +48,21 @@ class ChainDeploymentHandle(typing.NamedTuple):
...
+    current_hash: str
+    current_signature: str
+
+
+class TrussPatches(typing.NamedTuple):
+    django_patch_state: PatchState
+    container_patch_state: PatchState
+
+
+class TrussWatchState(typing.NamedTuple):
```
Comment (squidarth) [inline truss/remote/baseten/core.py:61]: nit: seems to me that we could model these types in a similar way to the backend

```
class TrussWatchState
     is_container_built_from_push: bool
    container_patch_state: PatchState
    ...
```

why do things differently?

Comment (rcano-baseten) [reply truss/remote/baseten/core.py:61]: this way plays nicer with mypy - you ensure that both of pieces exist which more accurate reflects the different states

{bool = False, empty state} vs {bool = True, present state} I can update the server side response to just be this - lmk what you think

File: truss/remote/baseten/remote.py (hunk 2 above)
Comment (rcano-baseten) [reply truss/remote/baseten/remote.py:377]: specifically that django has patches that are unapplied to the truss container

Comment (rcano-baseten) [reply truss/remote/baseten/remote.py:377]: yeah i think so

**Review by squidarth — APPROVED** [review]

**Pull request merged**